crates/core/tedge_api/src/mqtt_topics.rs

```diff
@@ -628,6 +631,18 @@ pub enum ChannelFilter {
     CommandMetadata(OperationType),
 }
 
+pub fn command_id_requester_with_timestamp(requester: &str) -> Result<String, error::Format> {
+    Ok(format!(
+        "{}-{}",
+        requester,
+        OffsetDateTime::now_utc().format(&format_description::well_known::Rfc3339)?,
```

Comment:
Is this date time precise enough to avoid collision if the mapper triggers two operations in a row? Not sure.

Hence, I will add a sequence number. For that you might have to change the API.

```
pub struct IdGenerator {
    prefix: String,
    next_id: usize, 
}

impl IdGenerator {
    pub fn new(prefix: &str) -> Self { ... }
    pub fn new_id(&mut self) -> String {
        // increment self.next_id
        // combine prefix + date + sequence into a unique id
    }
   pub fn is_generator_of(&self, &str) {
       // check the prefix
   }
}
```

Reply:
This is exemplary date in that format : `2023-10-30T19:58:26.649479831Z`  so I think for sequence of operations it should be more than enough to provide unique date. However, If we follow the solution above, how do we determine that counter should be incremented? Or by sequence number you mean randomly generated number?

Reply:
> This is exemplary date in that format : `2023-10-30T19:58:26.649479831Z`  so I think for sequence of operations it should be more than enough to provide unique date.

Using such a precision should indeed be enough (even if the nano figures is likely not reliable).

> However, If we follow the solution above, how do we determine that counter should be incremented? Or by sequence number you mean randomly generated number?

I mean a sequence number incremented each time a fresh identifier is requested.

- `c8y-mapper-2023-10-30T19:58:26-1`
- `c8y-mapper-2023-10-30T20:00:00-2`
- ...

Reply:
To clarify my point of view.

- I have no strong opinion about the precise string format used to generate used to generate command identifiers as long as they are unique and providing some hints about their provenance and age.
- I do think that wrapping the prefix in a `struct IdGenerator` equipped with methods to generate and check identifier with a given prefix is a better design compared to a bunch of functions and string constants. Using a struct leads to better encapsulation (I agree that here there not so much to abstract) and is easier to extend (say by adding later a sequence number).
- In too many places, the thin-edge code base started with constants such as `REQUESTER_NAME` leading to a code base that this is easy to break and hard to change. Instead of this constant, the mapper needs an id generator that encapsulates the prefix and the associated logic.

Reply:
I think nano seconds level is unique enough.